File: templates/wasm-node-assemblyscript/examples/string_nodes.ts
/**
 * String Nodes - Text manipulation utilities
 *
 * This example demonstrates string processing nodes including
 * case conversion, splitting, joining, and text analysis.
 */

import {
	type ExecutionInput,
	ExecutionResult,
	NodeDefinition,
	PinDefinition,
	jsonString,
	parseInput,
	serializeDefinition,
	serializeResult,
} from "../assembly/sdk";

// ============================================================================
// Uppercase Node
// ============================================================================

export function get_uppercase_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_uppercase_as";
	def.friendly_name = "To Uppercase (AS)";
	def.description = "Converts text to uppercase";
	def.category = "String/Transform";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input(
			"text",
			"Text",
			"Text to convert",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Result", "Uppercase text", "String"),
	);

	return serializeDefinition(def);
}

export function run_uppercase(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");

	const result = ExecutionResult.success();
	result.setOutput("result", jsonString(text.toUpperCase()));
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Lowercase Node
// ============================================================================

export function get_lowercase_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_lowercase_as";
	def.friendly_name = "To Lowercase (AS)";
	def.description = "Converts text to lowercase";
	def.category = "String/Transform";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input(
			"text",
			"Text",
			"Text to convert",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Result", "Lowercase text", "String"),
	);

	return serializeDefinition(def);
}

export function run_lowercase(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");

	const result = ExecutionResult.success();
	result.setOutput("result", jsonString(text.toLowerCase()));
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Trim Node
// ============================================================================

export function get_trim_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_trim_as";
	def.friendly_name = "Trim (AS)";
	def.description = "Removes leading and trailing whitespace";
	def.category = "String/Transform";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("text", "Text", "Text to trim", "String").withDefault(
			'""',
		),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Result", "Trimmed text", "String"),
	);

	return serializeDefinition(def);
}

export function run_trim(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");

	const result = ExecutionResult.success();
	result.setOutput("result", jsonString(text.trim()));
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// String Length Node
// ============================================================================

export function get_length_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_length_as";
	def.friendly_name = "String Length (AS)";
	def.description = "Returns the length of a string";
	def.category = "String/Analysis";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input(
			"text",
			"Text",
			"Text to measure",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("length", "Length", "Number of characters", "I64"),
	);
	def.addPin(
		PinDefinition.output(
			"is_empty",
			"Is Empty",
			"True if string is empty",
			"Bool",
		),
	);

	return serializeDefinition(def);
}

export function run_length(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");

	const result = ExecutionResult.success();
	result.setOutput("length", text.length.toString());
	result.setOutput("is_empty", (text.length == 0).toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Contains Node
// ============================================================================

export function get_contains_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_contains_as";
	def.friendly_name = "Contains (AS)";
	def.description = "Checks if text contains a substring";
	def.category = "String/Analysis";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input(
			"text",
			"Text",
			"Text to search in",
			"String",
		).withDefault('""'),
	);
	def.addPin(
		PinDefinition.input(
			"search",
			"Search",
			"Substring to find",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Found", "True if substring found", "Bool"),
	);

	return serializeDefinition(def);
}

export function run_contains(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");
	const search = getStringInput(input, "search", "");

	const result = ExecutionResult.success();
	result.setOutput("result", text.includes(search).toString());
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Replace Node
// ============================================================================

export function get_replace_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_replace_as";
	def.friendly_name = "Replace (AS)";
	def.description = "Replaces occurrences of a pattern";
	def.category = "String/Transform";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("text", "Text", "Original text", "String").withDefault(
			'""',
		),
	);
	def.addPin(
		PinDefinition.input(
			"find",
			"Find",
			"Pattern to find",
			"String",
		).withDefault('""'),
	);
	def.addPin(
		PinDefinition.input(
			"replace_with",
			"Replace With",
			"Replacement text",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Result", "Modified text", "String"),
	);

	return serializeDefinition(def);
}

export function run_replace(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const text = getStringInput(input, "text", "");
	const find = getStringInput(input, "find", "");
	const replaceWith = getStringInput(input, "replace_with", "");

	const result = ExecutionResult.success();
	result.setOutput("result", jsonString(text.replaceAll(find, replaceWith)));
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Concat Node
// ============================================================================

export function get_concat_definition(): i64 {
	const def = new NodeDefinition();
	def.name = "string_concat_as";
	def.friendly_name = "Concatenate (AS)";
	def.description = "Joins two strings together";
	def.category = "String/Transform";

	def.addPin(PinDefinition.input("exec", "Execute", "Trigger", "Exec"));
	def.addPin(
		PinDefinition.input("a", "A", "First string", "String").withDefault('""'),
	);
	def.addPin(
		PinDefinition.input("b", "B", "Second string", "String").withDefault('""'),
	);
	def.addPin(
		PinDefinition.input(
			"separator",
			"Separator",
			"Text between strings",
			"String",
		).withDefault('""'),
	);
	def.addPin(PinDefinition.output("exec_out", "Done", "Complete", "Exec"));
	def.addPin(
		PinDefinition.output("result", "Result", "Combined string", "String"),
	);

	return serializeDefinition(def);
}

export function run_concat(ptr: i32, len: i32): i64 {
	const input = parseInput(ptr, len);
	const a = getStringInput(input, "a", "");
	const b = getStringInput(input, "b", "");
	const separator = getStringInput(input, "separator", "");

	const result = ExecutionResult.success();
	result.setOutput("result", jsonString(a + separator + b));
	result.activateExec("exec_out");

	return serializeResult(result);
}

// ============================================================================
// Helpers
// ============================================================================

function getStringInput(
	input: ExecutionInput,
	name: string,
	defaultValue: string,
): string {
	if (input.inputs.has(name)) {
		const value = input.inputs.get(name);
		if (value.startsWith('"') && value.endsWith('"')) {
			return value.slice(1, value.length - 1);
		}
		return value;
	}
	return defaultValue;
}

File: templates/wasm-node-rust/examples/string_nodes.rs
//! String Nodes - Text manipulation utilities
//!
//! This example demonstrates string processing nodes including
//! case conversion, splitting, joining, and text analysis.

use flow_like_wasm_sdk::*;

package! {
    nodes: [
        {
            name: "string_uppercase",
            friendly_name: "To Uppercase",
            description: "Converts text to uppercase",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                text: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
            },
        },
        {
            name: "string_lowercase",
            friendly_name: "To Lowercase",
            description: "Converts text to lowercase",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                text: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
            },
        },
        {
            name: "string_trim",
            friendly_name: "Trim",
            description: "Removes leading and trailing whitespace",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                text: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
            },
        },
        {
            name: "string_length",
            friendly_name: "String Length",
            description: "Returns the length of a string",
            category: "String/Analysis",
            inputs: {
                exec: Exec,
                text: String = "",
            },
            outputs: {
                exec_out: Exec,
                length: I64,
                is_empty: Bool,
            },
        },
        {
            name: "string_contains",
            friendly_name: "Contains",
            description: "Checks if text contains a substring",
            category: "String/Analysis",
            inputs: {
                exec: Exec,
                text: String = "",
                search: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: Bool,
            },
        },
        {
            name: "string_replace",
            friendly_name: "Replace",
            description: "Replaces occurrences of a pattern",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                text: String = "",
                find: String = "",
                replace_with: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
                count: I64,
            },
        },
        {
            name: "string_concat",
            friendly_name: "Concatenate",
            description: "Joins multiple strings together",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                a: String = "",
                b: String = "",
                separator: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
            },
        },
        {
            name: "string_reverse",
            friendly_name: "Reverse",
            description: "Reverses the characters in a string",
            category: "String/Transform",
            inputs: {
                exec: Exec,
                text: String = "",
            },
            outputs: {
                exec_out: Exec,
                result: String,
            },
        }
    ]
}

run_package!(run_node);

fn run_node(node_name: &str, mut ctx: Context) -> ExecutionResult {
    match node_name {
        "string_uppercase" => {
            let text = ctx.get_string("text").unwrap_or_default();
            ctx.set_output("result", text.to_uppercase());
            ctx.success()
        }
        "string_lowercase" => {
            let text = ctx.get_string("text").unwrap_or_default();
            ctx.set_output("result", text.to_lowercase());
            ctx.success()
        }
        "string_trim" => {
            let text = ctx.get_string("text").unwrap_or_default();
            ctx.set_output("result", text.trim().to_string());
            ctx.success()
        }
        "string_length" => {
            let text = ctx.get_string("text").unwrap_or_default();
            ctx.set_output("length", text.len() as i64);
            ctx.set_output("is_empty", text.is_empty());
            ctx.success()
        }
        "string_contains" => {
            let text = ctx.get_string("text").unwrap_or_default();
            let search = ctx.get_string("search").unwrap_or_default();
            ctx.set_output("result", text.contains(&search));
            ctx.success()
        }
        "string_replace" => {
            let text = ctx.get_string("text").unwrap_or_default();
            let find = ctx.get_string("find").unwrap_or_default();
            let replace_with = ctx.get_string("replace_with").unwrap_or_default();
            let count = text.matches(&find).count() as i64;
            let result = text.replace(&find, &replace_with);
            ctx.set_output("result", result);
            ctx.set_output("count", count);
            ctx.success()
        }
        "string_concat" => {
            let a = ctx.get_string("a").unwrap_or_default();
            let b = ctx.get_string("b").unwrap_or_default();
            let separator = ctx.get_string("separator").unwrap_or_default();
            let result = if separator.is_empty() {
                format!("{}{}", a, b)
            } else {
                format!("{}{}{}", a, separator, b)
            };
            ctx.set_output("result", result);
            ctx.success()
        }
        "string_reverse" => {
            let text = ctx.get_string("text").unwrap_or_default();
            ctx.set_output("result", text.chars().rev().collect::<String>());
            ctx.success()
        }
        _ => ctx.fail(format!("Unknown node: {}", node_name)),
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_uppercase() {
        assert_eq!("hello".to_uppercase(), "HELLO");
    }

    #[test]
    fn test_reverse() {
        let s = "hello";
        let reversed: String = s.chars().rev().collect();
        assert_eq!(reversed, "olleh");
    }

    #[test]
    fn test_contains() {
        assert!("hello world".contains("world"));
        assert!(!"hello world".contains("foo"));
    }
}
